### PR TITLE
feat: display trial restarts [DET-7347]

### PR DESCRIPTION
**Pull request by hkang1**
## Description

Expose number of times a trial restarted for both single and multi-trial experiments.

Single trial experiments show the restarts in the experiment header as such:
Screen Shot 2022-05-19 at 3 09 42 PM

Multi-trial experiments show the restarts in the trials table column as such:
Screen Shot 2022-05-19 at 2 10 20 PM

## Test Plan

Restarts can be done manually (pause / restart)
- [ ] verify that the restarting increments the restarts counter
- [ ] verify that single trial experiments show the restart count in the experiment header.
- [ ] verify that the multi-trial experiments show the restart count in the trials table

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by stoksc — APPROVED** [review]
lgtm

File: webui/react/src/pages/ExperimentDetails/ExperimentTrials.tsx [hunk 1]
```diff
@@ -101,6 +101,13 @@ const ExperimentTrials: React.FC<Props> = ({ experiment }: Props) => {
       </Link>
     );
 
+    const restartsRenderer = (_: string, record: TrialItem): React.ReactNode => {
```
Comment (hamidzr) [inline webui/react/src/pages/ExperimentDetails/ExperimentTrials.tsx:104]: suggestion. only render if there has been a restart? these should be rare cases.

- also if it only shows up conditionally it'd be more eye-grabbing which is probably a reason behind why they asked for the feature.

File: webui/react/src/pages/ExperimentDetails/ExperimentTrials.tsx [hunk 2]
```diff
@@ -101,6 +101,13 @@ const ExperimentTrials: React.FC<Props> = ({ experiment }: Props) => {
...
+      const maxRestarts = experiment.config.maxRestarts ?? 0;
+      return (
+        <>{record.restarts}{maxRestarts ? `/${maxRestarts}` : ''}</>
```
Comment (hamidzr) [inline webui/react/src/pages/ExperimentDetails/ExperimentTrials.tsx:107]: @stoksc isn't max restarts always capped? can the user opt to make it go on forever?

File: webui/react/src/services/api-ts-sdk/api.ts [hunk 3]
```diff
@@ -17795,7 +17802,7 @@ export class TrialsApi extends BaseAPI {
...
+    public getExperimentTrials(experimentId: number, sortBy?: 'SORT_BY_UNSPECIFIED' | 'SORT_BY_ID' | 'SORT_BY_START_TIME' | 'SORT_BY_END_TIME' | 'SORT_BY_STATE' | 'SORT_BY_BEST_VALIDATION_METRIC' | 'SORT_BY_LATEST_VALIDATION_METRIC' | 'SORT_BY_BATCHES_PROCESSED' | 'SORT_BY_DURATION' | 'SORT_BY_RESTARTS', orderBy?: 'ORDER_BY_UNSPECIFIED' | 'ORDER_BY_ASC' | 'ORDER_BY_DESC', offset?: number, limit?: number, states?: Array<'STATE_UNSPECIFIED' | 'STATE_ACTIVE' | 'STATE_PAUSED' | 'STATE_STOPPING_COMPLETED' | 'STATE_STOPPING_CANCELED' | 'STATE_STOPPING_ERROR' | 'STATE_COMPLETED' | 'STATE_CANCELED' | 'STATE_ERROR' | 'STATE_DELETED' | 'STATE_DELETING' | 'STATE_DELETE_FAILED' | 'STATE_STOPPING_KILLED'>, options?: any) {
```
Comment (hamidzr) [inline webui/react/src/services/api-ts-sdk/api.ts:17805]: unrelated: I wish they did this a better way. although this is not the generator's fault. the swagger spec (or the spec generator) doesn't link the parameter type to enum definition it just reiterates the values..

**Review by hamidzr — COMMENTED** [review]
A minor issue and a question otherwise the code changes look good to me. 

Screen Shot 2022-05-19 at 02 11 07

Whether we're displaying these in the best place for the user or not I don't have enough context on and would depend on the user stories among other things. eg would they want to see the restarts when looking at a single trial from a multi-trial experiment?

interesting cases IMO
- trials/experiments that have recovered with some restarts
- trials that have not recovered and have exhausted all of the restart budget

**Review by trentwatt — COMMENTED** [review]
hMMM....

basically: db says restarts: 6 (max restarts is 5: is it counting the initial start as a restart?)

API says restarts: 0


## trial is definitely restarting (several times):

image



## but restarts says zero in the table:

image

## and in the API:

Screen Shot 2022-05-19 at 3 11 43 AM


## But.... 

image

File: webui/react/src/pages/ExperimentDetails/ExperimentTrials.tsx (hunk 2 above)
Comment (stoksc) [reply webui/react/src/pages/ExperimentDetails/ExperimentTrials.tsx:107]: user opt makes it (0, max_int)

File: webui/react/src/pages/ExperimentDetails/ExperimentTrials.tsx (hunk 1 above)
Comment (alexjdetermined) [reply webui/react/src/pages/ExperimentDetails/ExperimentTrials.tsx:104]: Agreed. This makes a lot of sense.

File: webui/react/src/services/api-ts-sdk/api.ts (hunk 3 above)
Comment (hkang1) [reply webui/react/src/services/api-ts-sdk/api.ts:17805]: ➕

File: master/static/srv/proto_get_trials_plus.sql [hunk 4]
```diff
@@ -177,7 +178,8 @@ SELECT
         UNION ALL
         SELECT jsonb_each(resources) FROM checkpoints_new_view c WHERE c.trial_id = t.id
     ) r
-  ) AS total_checkpoint_size
+  ) AS total_checkpoint_size,
+  LEAST(t.restarts, max_restarts) as restarts
```
Comment (stoksc) [inline master/static/srv/proto_get_trials_plus.sql:182]: please drop a comment explaining why LEAST is needed (I get it, we save t.restarts++ even when we don't restart the last time, and changing that doesn't help us for old trials, so just make it clear)

**Review by stoksc — APPROVED** [review]
lgtm modulo final comment

**Pull request merged**